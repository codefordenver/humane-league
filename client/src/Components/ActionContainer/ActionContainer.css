.ActionContainer {
  align-items: center;
  display: flex;
  flex-direction: column;
  padding-bottom: 50px; }
  .ActionContainer h1 {
    color: white;
    font-weight: normal;
    font-size: 36px; }
