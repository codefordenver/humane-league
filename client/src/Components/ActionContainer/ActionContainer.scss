.ActionContainer {
  align-items: center;
  display: flex;
  flex-direction: column;
  padding-bottom: 50px;

  h1 {
    color: white;
    font-weight: normal;
    font-size: 36px;
  }
}