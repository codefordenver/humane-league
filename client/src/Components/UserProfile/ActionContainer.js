import React from 'react';
import ActionCard from './ActionCard';

const ActionContainer = () => {
  return (
    <div>
      I am an ActionContainer
      <ActionCard />
    </div>
  );
}

export default ActionContainer;
