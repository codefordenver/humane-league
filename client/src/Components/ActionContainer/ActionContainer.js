/* eslint-disable no-undef */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import TwitterCard from '../ActionCard/TwitterCard';
import FacebookCard from '../ActionCard/FacebookCard';
import EmailCard from '../ActionCard/EmailCard';
import PhoneCard from '../ActionCard/PhoneCard';
import './ActionContainer.css';
import '../ActionCard/ActionCard.css';
import { 
  getTwitterActions, 
  getFacebookActions, 
  getEmailActions, 
  getPhoneActions,
  getCompletedActions } from '../../utils/apiCalls';

export class ActionContainer extends Component {
  constructor() {
    super();
    this.state = {
      loading: true,
      userPreferences: {},
      twitter_actions: [],
      facebook_actions: [],
      email_actions: [],
      phone_actions: []
    };
  }
  async componentDidMount() {
    const { twitter_actions, facebook_actions, email_actions, phone_actions, id, id_token} = this.props.user;
    await this.setState({ userPreferences: { twitter_actions, facebook_actions, email_actions, phone_actions} });

    const completedActions = await getCompletedActions(id, id_token);
    const actions = {
      twitter_actions: this.state.userPreferences.twitter_actions ? 
        await getTwitterActions() : [],
      facebook_actions: this.state.userPreferences.facebook_actions ?
        await getFacebookActions() : [],
      email_actions: this.state.userPreferences.email_actions ? 
        await getEmailActions() : [],
      phone_actions: this.state.userPreferences.phone_actions ?
        await getPhoneActions() : []
    };

    completedActions.forEach( action => {
      actions[action.action_type] = actions[action.action_type].filter(act => act.id != action.action_id);
    });

    Object.keys(actions).forEach(actionType => {
      actions[actionType] = actions[actionType].filter(action => action.enabled === true);
      actions[actionType].sort((a, b) => a.created_at > b.created_at);
    });
    console.log(this.state);
    await this.setState(Object.assign(actions, { loading: false }));
    console.log(this.state);
  }

  removeCompleted = (actionType, action) => {
    const index = this.state[actionType].indexOf(action);
    const actions = this.state[actionType];
    actions[index] = { id: action.id, completed: true, title: 'Thank You!', description: `ðŸŽ‰  For taking action: ${action.title}  ðŸŽ‰`};
    setTimeout(() => this.setState({ [actionType]: actions }), 1000);
  }

  render() {
    const { twitter_actions, facebook_actions, email_actions, phone_actions } = this.state;
    const { user } = this.props;

    const twitter = twitter_actions.length ? <TwitterCard action={twitter_actions[0]} user={user} removeCompleted={this.removeCompleted}/> : null;
    const facebook = facebook_actions.length ? <FacebookCard action={facebook_actions[0]} user={user} removeCompleted={this.removeCompleted}/> : null;
    const email = email_actions.length ? <EmailCard action={email_actions[0]} user={user} removeCompleted={this.removeCompleted}/> : null;
    const phone = phone_actions.length ? <PhoneCard action={phone_actions[0]} user={user} removeCompleted={this.removeCompleted}/> : null;

    const noActions = twitter_actions.length === 0 && facebook_actions.length === 0 && email_actions.length === 0 && phone_actions.length === 0 && !this.state.loading;
    const noActionsMessage = noActions ? <div className="no-actions"><p>You have completed all of our actions!</p><p>ðŸŽ‰  Great job!  ðŸŽ‰</p><p> Please check back again soon for new actions. In the meantime, check out <a className="thl-twitter-link" href="https://twitter.com/thehumaneleague" target="_blank">our Twitter feed</a> to see what we're up to.</p></div> : null;
    const loadingMessage = this.state.loading ? <p className="no-actions">Loading...</p> : null;


    return (
      <div className="ActionContainer">
        <h1>FAN ACTION ALERTS</h1>
        {twitter}
        {facebook}
        {email}
        {phone}
        {loadingMessage}
        {noActionsMessage}
      </div>
    );
  }
}

const mapStateToProps = store => ({
  user: store.User
});

export default connect(mapStateToProps, null)(ActionContainer);

ActionContainer.propTypes = {
  user: PropTypes.object
};
