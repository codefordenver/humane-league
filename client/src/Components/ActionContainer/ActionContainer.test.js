import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { ActionContainer } from './ActionContainer';

describe('ActionContainer component tests', () => {
  let renderedActionContainer;
  let defaultState;
  let mockUser;
  let mockAction;
  beforeEach(() => {
    mockUser = { twitter_actions: false, facebook_actions: false, email_actions: false, phone_actions: false };
    renderedActionContainer = shallow( <ActionContainer user={mockUser}/> );
    defaultState = {
      userPreferences: {
        ...mockUser
      },
      twitter: [],
      facebook: [],
      email: [],
      phone: []
    };
    mockAction = { title: 'action title' };

    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => 
        Promise.resolve(
          { results: {
              // twitter: [mockAction, mockAction],
              // facebook: [mockAction, mockAction],
              // email: [mockAction, mockAction],
              // phone: [mockAction, mockAction]
            }
          }
        )   
    }));

  })
  it('renders without crashing', () => {
    expect(renderedActionContainer).toBeDefined();
    expect(renderedActionContainer.state()).toEqual(defaultState);
  });
  it('matches SnapShot', () => {
    expect(renderedActionContainer).toMatchSnapshot();
  })
  
})