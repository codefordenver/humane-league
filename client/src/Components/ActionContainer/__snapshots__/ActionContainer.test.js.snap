// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionContainer component tests matches SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ActionContainer
    user={
        Object {
            "email_actions": false,
            "facebook_actions": false,
            "phone_actions": false,
            "twitter_actions": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          FAN ACTION ALERTS
</h1>,
        null,
        null,
        null,
        null,
        <p
          className="no-actions"
>
          Loading...
</p>,
        null,
      ],
      "className": "ActionContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "FAN ACTION ALERTS",
        },
        "ref": null,
        "rendered": "FAN ACTION ALERTS",
        "type": "h1",
      },
      null,
      null,
      null,
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Loading...",
          "className": "no-actions",
        },
        "ref": null,
        "rendered": "Loading...",
        "type": "p",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            FAN ACTION ALERTS
</h1>,
          null,
          null,
          null,
          null,
          <p
            className="no-actions"
>
            Loading...
</p>,
          null,
        ],
        "className": "ActionContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "FAN ACTION ALERTS",
          },
          "ref": null,
          "rendered": "FAN ACTION ALERTS",
          "type": "h1",
        },
        null,
        null,
        null,
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Loading...",
            "className": "no-actions",
          },
          "ref": null,
          "rendered": "Loading...",
          "type": "p",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
